File: src/main/java/alfheim/common/item/equipment/armor/elvoruim/ItemElvoriumHelmet.java
package alfheim.common.item.equipment.armor.elvoruim;

import static org.lwjgl.opengl.GL11.*;

import java.util.List;

import cpw.mods.fml.common.eventhandler.SubscribeEvent;
import cpw.mods.fml.relauncher.Side;
import cpw.mods.fml.relauncher.SideOnly;
import net.minecraft.client.Minecraft;
import net.minecraft.client.renderer.ItemRenderer;
import net.minecraft.client.renderer.Tessellator;
import net.minecraft.client.renderer.texture.IIconRegister;
import net.minecraft.client.renderer.texture.TextureMap;
import net.minecraft.entity.Entity;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.ItemStack;
import net.minecraft.potion.Potion;
import net.minecraft.potion.PotionEffect;
import net.minecraft.util.IIcon;
import net.minecraft.util.StatCollector;
import net.minecraft.world.World;
import net.minecraftforge.client.event.RenderPlayerEvent;
import net.minecraftforge.common.MinecraftForge;
import net.minecraftforge.event.entity.living.LivingHurtEvent;
import vazkii.botania.api.item.IAncientWillContainer;
import vazkii.botania.api.mana.IManaGivingItem;
import vazkii.botania.api.mana.ManaItemHandler;
import vazkii.botania.client.core.helper.IconHelper;
import vazkii.botania.common.core.helper.ItemNBTHelper;

public class ItemElvoriumHelmet extends ItemElvoriumArmor implements IAncientWillContainer, IManaGivingItem {

	private static final String TAG_ANCIENT_WILL = "AncientWill";
	static IIcon willIcon;

	public ItemElvoriumHelmet() {
		this("ElvoriumHelmet");
		MinecraftForge.EVENT_BUS.register(this);
	}

	public ItemElvoriumHelmet(String name) {
		super(0, name);
	}

	@Override
	@SideOnly(Side.CLIENT)
	public void registerIcons(IIconRegister reg) {
		super.registerIcons(reg);
		willIcon = IconHelper.forName(reg, "willFlame");
	}

	@Override
	public void onArmorTick(World world, EntityPlayer player, ItemStack stack) {
		super.onArmorTick(world, player, stack);
		if(hasArmorSet(player)) {
			int food = player.getFoodStats().getFoodLevel();
			if(food > 0 && food < 18 && player.shouldHeal() && player.ticksExisted % 80 == 0)
				player.heal(1F);
			ManaItemHandler.dispatchManaExact(stack, player, 1, true);
		}
	}

	@Override
	public void addAncientWill(ItemStack stack, int will) {
		ItemNBTHelper.setBoolean(stack, TAG_ANCIENT_WILL + will, true);
	}

	@Override
	public boolean hasAncientWill(ItemStack stack, int will) {
		return hasAncientWill_(stack, will);
	}

	public static boolean hasAncientWill_(ItemStack stack, int will) {
		return ItemNBTHelper.getBoolean(stack, TAG_ANCIENT_WILL + will, false);
	}

	@Override
	@SideOnly(Side.CLIENT)
	public void addArmorSetDescription(ItemStack stack, List<String> list) {
		super.addArmorSetDescription(stack, list);
		for(int i = 0; i < 6; i++)
			if(hasAncientWill(stack, i))
				addStringToTooltip(StatCollector.translateToLocal("botania.armorset.will" + i + ".desc"), list);
	}

	public static boolean hasAnyWill(ItemStack stack) {
		for(int i = 0; i < 6; i++)
			if(hasAncientWill_(stack, i))
				return true;

		return false;
	}

	@SideOnly(Side.CLIENT)
	public static void renderOnPlayer(ItemStack stack, RenderPlayerEvent event) {
		if(hasAnyWill(stack) && !((ItemElvoriumArmor) stack.getItem()).hasPhantomInk(stack)) {
			glPushMatrix();
			float f = willIcon.getMinU();
			float f1 = willIcon.getMaxU();
			float f2 = willIcon.getMinV();
			float f3 = willIcon.getMaxV();
			vazkii.botania.api.item.IBaubleRender.Helper.translateToHeadLevel(event.entityPlayer);
			Minecraft.getMinecraft().renderEngine.bindTexture(TextureMap.locationItemsTexture);
			glRotated(90, 0, 1, 0);
			glRotated(180, 1, 0, 0);
			glTranslated(-0.26, 0.15, -0.39);
			glScaled(0.5, 0.5, 0.5);
			ItemRenderer.renderItemIn2D(Tessellator.instance, f1, f2, f, f3, willIcon.getIconWidth(), willIcon.getIconHeight(), 1F / 16F);
			glPopMatrix();
		}
	}

	@SubscribeEvent
	public void onEntityAttacked(LivingHurtEvent event) {
		Entity attacker = event.source.getEntity();
		if(attacker instanceof EntityPlayer) {
			EntityPlayer player = (EntityPlayer) attacker;
			if(hasArmorSet(player)) {
				boolean crit = player.fallDistance > 0.0F && !player.onGround && !player.isOnLadder() && !player.isInWater() && !player.isPotionActive(Potion.blindness) && player.ridingEntity == null;
				ItemStack stack = player.inventory.armorItemInSlot(3);
				if(crit && stack != null && stack.getItem() instanceof ItemElvoriumHelmet) {
					boolean ahrim = hasAncientWill(stack, 0);
					boolean dharok = hasAncientWill(stack, 1);
					boolean guthan = hasAncientWill(stack, 2);
					boolean torag = hasAncientWill(stack, 3);
					boolean verac = hasAncientWill(stack, 4);
					boolean karil = hasAncientWill(stack, 5);

					if(ahrim)
						event.entityLiving.addPotionEffect(new PotionEffect(Potion.weakness.id, 20, 1));
					if(dharok)
						event.ammount *= 1F + (1F - player.getHealth() / player.getMaxHealth()) * 0.5F;
					if(guthan)
						player.heal(event.ammount * 0.25F);
					if(torag)
						event.entityLiving.addPotionEffect(new PotionEffect(Potion.moveSlowdown.id, 60, 1));
					if(verac)
						event.source.setDamageBypassesArmor();
					if(karil)
						event.entityLiving.addPotionEffect(new PotionEffect(Potion.wither.id, 60, 1));
				}
			}
		}
	}
	
	@SubscribeEvent
	@SideOnly(Side.CLIENT)
	public void onPlayerRender(RenderPlayerEvent.Specials.Post event) {
		if(event.entityLiving.getActivePotionEffect(Potion.invisibility) != null)
			return;

		EntityPlayer player = event.entityPlayer;

		float yaw = player.prevRotationYawHead + (player.rotationYawHead - player.prevRotationYawHead) * event.partialRenderTick;
		float yawOffset = player.prevRenderYawOffset + (player.renderYawOffset - player.prevRenderYawOffset) * event.partialRenderTick;
		float pitch = player.prevRotationPitch + (player.rotationPitch - player.prevRotationPitch) * event.partialRenderTick;

		glPushMatrix();
		glRotatef(yawOffset, 0, -1, 0);
		glRotatef(yaw - 270, 0, 1, 0);
		glRotatef(pitch, 0, 0, 1);

		ItemStack helm = player.inventory.armorItemInSlot(3);
		if(helm != null && helm.getItem() instanceof ItemElvoriumHelmet)
			ItemElvoriumHelmet.renderOnPlayer(helm, event);

		glPopMatrix();
	}
}

File: src/main/java/alfheim/common/item/equipment/armor/elvoruim/ItemElvoriumHelmet.kt
package alfheim.common.item.equipment.armor.elvoruim

import alexsocol.asjlib.*
import alfheim.common.core.handler.AlfheimConfigHandler
import cpw.mods.fml.common.eventhandler.SubscribeEvent
import cpw.mods.fml.relauncher.*
import net.minecraft.client.renderer.*
import net.minecraft.client.renderer.texture.*
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.item.ItemStack
import net.minecraft.potion.*
import net.minecraft.util.*
import net.minecraft.world.World
import net.minecraftforge.client.event.RenderPlayerEvent
import net.minecraftforge.common.MinecraftForge
import net.minecraftforge.event.entity.living.LivingHurtEvent
import org.lwjgl.opengl.GL11.*
import vazkii.botania.api.item.IAncientWillContainer
import vazkii.botania.api.mana.IManaGivingItem
import vazkii.botania.client.core.helper.IconHelper
import vazkii.botania.common.core.helper.ItemNBTHelper

open class ItemElvoriumHelmet(name: String): ItemElvoriumArmor(0, name), IAncientWillContainer, IManaGivingItem {
	
	constructor(): this("ElvoriumHelmet") {
		MinecraftForge.EVENT_BUS.register(this)
	}
	
	@SideOnly(Side.CLIENT)
	override fun registerIcons(reg: IIconRegister) {
		super.registerIcons(reg)
		willIcon = IconHelper.forName(reg, "willFlame")
	}
	
	override fun onArmorTick(world: World, player: EntityPlayer, stack: ItemStack) {
		super.onArmorTick(world, player, stack)
		if (hasArmorSet(player)) {
			val food = player.foodStats.foodLevel
			if (food in 1..17 && player.shouldHeal() && player.ticksExisted % 80 == 0)
				player.heal(1f)
			
//			ManaItemHandler.dispatchManaExact(stack, player, 2, true)
		}
	}
	
	override fun addAncientWill(stack: ItemStack, will: Int) {
		ItemNBTHelper.setBoolean(stack, TAG_ANCIENT_WILL + will, true)
	}
	
	override fun hasAncientWill(stack: ItemStack?, will: Int): Boolean {
		return hasAncientWill_(stack, will)
	}
	
	@SideOnly(Side.CLIENT)
	override fun addArmorSetDescription(stack: ItemStack?, list: List<String>) {
		super.addArmorSetDescription(stack, list)
		
		for (i in 0..6)
			if (hasAncientWill(stack, i))
				addStringToTooltip(StatCollector.translateToLocal("botania.armorset.will$i.desc"), list)
	}
	
	@SubscribeEvent
	fun onEntityAttacked(event: LivingHurtEvent) {
		val attacker = event.source.entity as? EntityPlayer ?: return
		
		if (!hasArmorSet(attacker)) return
		
		val stack = attacker.inventory.armorItemInSlot(3) ?: return
		val crit = attacker.fallDistance > 0f && !attacker.onGround && !attacker.isOnLadder && !attacker.isInWater && !attacker.isPotionActive(Potion.blindness) && attacker.ridingEntity == null
		
		if (crit && stack.item is ItemElvoriumHelmet) {
			if (hasAncientWill(stack, 0))
				event.entityLiving.addPotionEffect(PotionEffect(Potion.weakness.id, 20, 1))
			
			if (hasAncientWill(stack, 1))
				event.ammount *= 1f + (1f - attacker.health / attacker.maxHealth) * 0.5f
			
			if (hasAncientWill(stack, 2))
				attacker.heal(event.ammount * 0.25f)
			
			if (hasAncientWill(stack, 3))
				event.entityLiving.addPotionEffect(PotionEffect(Potion.moveSlowdown.id, 60, 1))
			
			if (hasAncientWill(stack, 4))
				event.source.setDamageBypassesArmor()
			
			if (hasAncientWill(stack, 5))
				event.entityLiving.addPotionEffect(PotionEffect(Potion.wither.id, 60, 1))
			
			if (hasAncientWill(stack, 6)) {
				event.entityLiving.addPotionEffect(PotionEffect(Potion.blindness.id, 60, 1))
				event.entityLiving.addPotionEffect(PotionEffect(AlfheimConfigHandler.potionIDEternity, 60, 1))
			}
		}
	}
	
	@SubscribeEvent
	@SideOnly(Side.CLIENT)
	fun onPlayerRender(event: RenderPlayerEvent.Specials.Post) {
		if (event.entityLiving.getActivePotionEffect(Potion.invisibility.id) != null)
			return
		
		val player = event.entityPlayer
		
		val yaw = player.prevRotationYawHead + (player.rotationYawHead - player.prevRotationYawHead) * event.partialRenderTick
		val yawOffset = player.prevRenderYawOffset + (player.renderYawOffset - player.prevRenderYawOffset) * event.partialRenderTick
		val pitch = player.prevRotationPitch + (player.rotationPitch - player.prevRotationPitch) * event.partialRenderTick
		
		glPushMatrix()
		glRotatef(yawOffset, 0f, -1f, 0f)
		glRotatef(yaw - 270, 0f, 1f, 0f)
		glRotatef(pitch, 0f, 0f, 1f)
		
		val helm = player.inventory.armorItemInSlot(3)
		if (helm != null && helm.item is ItemElvoriumHelmet)
			renderOnPlayer(helm, event)
		
		glPopMatrix()
	}
	
	companion object {
		
		private const val TAG_ANCIENT_WILL = "AncientWill"
		internal lateinit var willIcon: IIcon
		
		fun hasAncientWill_(stack: ItemStack?, will: Int): Boolean {
			return ItemNBTHelper.getBoolean(stack, TAG_ANCIENT_WILL + will, false)
		}
		
		fun hasAnyWill(stack: ItemStack): Boolean {
			for (i in 0..6)
				if (hasAncientWill_(stack, i))
					return true
			
			return false
		}
		
		@SideOnly(Side.CLIENT)
		fun renderOnPlayer(stack: ItemStack, event: RenderPlayerEvent) {
			if (hasAnyWill(stack) && !(stack.item as ItemElvoriumArmor).hasPhantomInk(stack)) {
				glPushMatrix()
				val f = willIcon.minU
				val f1 = willIcon.maxU
				val f2 = willIcon.minV
				val f3 = willIcon.maxV
				vazkii.botania.api.item.IBaubleRender.Helper.translateToHeadLevel(event.entityPlayer)
				mc.renderEngine.bindTexture(TextureMap.locationItemsTexture)
				glRotatef(90f, 0f, 1f, 0f)
				glRotatef(180f, 1f, 0f, 0f)
				glTranslated(-0.26, 0.15, -0.32)
				glScaled(0.5)
				ItemRenderer.renderItemIn2D(Tessellator.instance, f1, f2, f, f3, willIcon.iconWidth, willIcon.iconHeight, 1f / 16f)
				glPopMatrix()
			}
		}
	}
}
